### PR TITLE
fix JS optimizer fail on ES6 spread operator

### LINKED ISSUE #12543: Acorn optimizer - Cannot read property 'type' of undefined
My build is failing with the following error:

```
building:DEBUG: running cleanup on shell code: JSDCE minifyWhitespace
/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:91
    if (!(node.type in cs)) {
               ^

TypeError: Cannot read property 'type' of undefined
    at c (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:91:16)
    at /usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:395:11
    at Array.forEach (<anonymous>)
    at ObjectExpression (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:394:25)
    at c (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:96:20)
    at Object.VariableDeclarator (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:390:24)
    at c (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:96:20)
    at recursiveWalk (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:98:5)
    at /usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:93:9
    at maybeChild (/usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:51:7)
em++: error: '/usr/local/emsdk/node/12.18.1_64bit/bin/node /usr/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp/dwr.wasm.js.pp.js JSDCE minifyWhitespace --closureFriendly' failed (1)
cache:DEBUG: PID 110 released multiprocess file lock to Emscripten cache at /usr/local/emsdk/upstream/emscripten/cache/wasm
```

Here's the command I'm running:

```
/usr/local/emsdk/upstream/emscripten/em++.py -Oz --closure 1 -DSK_RELEASE -DGR_GL_CHECK_ALLOC_WITH_GET_ERROR=0 -I third_party/skia -I third_party/skia/third_party/icu -I third_party/skia/third_party/externals/icu/source/common/ -I include -DSK_DISABLE_AAA -DSK_FORCE_8_BYTE_ALIGNMENT -lGL -DSK_SUPPORT_GPU=1 -DSK_GL -DSK_DISABLE_LEGACY_SHADERCONTEXT -s USE_WEBGL2=1 -std=c++17 --bind --pre-js src/js/pre.js --pre-js src/js/interface.js src/Canvas.cpp src/Rectangle.cpp src/RenderList.cpp src/RenderObject.cpp src/RenderObjectContainer.cpp src/Shape.cpp src/Text.cpp src/main.cpp /root/design-wasm-renderer/third_party/skia/out/lib/libskparagraph.a /root/design-wasm-renderer/third_party/skia/out/lib/libskshaper.a /root/design-wasm-renderer/third_party/skia/out/lib/libharfbuzz.a /root/design-wasm-renderer/third_party/skia/out/lib/libicu.a /root/design-wasm-renderer/third_party/skia/out/lib/libskia.a /root/design-wasm-renderer/third_party/skia/modules/canvaskit/fonts/NotoMono-Regular.ttf.cpp -s EXPORT_NAME='Module' -s ALLOW_MEMORY_GROWTH=1 -s INITIAL_MEMORY=128MB -s NO_EXIT_RUNTIME=1 -s STRICT=1 -s WASM=1 -s USE_BOOST_HEADERS=1 -s LLD_REPORT_UNDEFINED -s ASSERTIONS=1 -s SAFE_HEAP=1 -Wl,-error-limit=0 -o out/dwr.js
```

version: 2.0.7

### PR DESCRIPTION
**Pull request by pavelsavara**
fixes https://github.com/emscripten-core/emscripten/issues/12543

### CODE REVIEW

**Comment by sbc100** [comment]
Would be good to add a test somewhere.  @kripken probably knows where the best place might be

**Review by kripken — APPROVED** [review]
lgtm too, thanks! For a test, please add something minimal here:

https://github.com/emscripten-core/emscripten/blob/main/test/optimizer/JSDCE.js

(expected output goes here: https://github.com/emscripten-core/emscripten/blob/main/test/optimizer/JSDCE-output.js )

**Comment by pavelsavara** [comment]
I don't have the tests running on my local box, so I will make test iterations on the CI. Sorry for the noise.

**Comment by kripken** [comment]
Running locally, this is the diff from `--rebaseline`:
```diff
diff --git a/test/optimizer/JSDCE-output.js b/test/optimizer/JSDCE-output.js
index 89c4957f7..480f0396b 100644
--- a/test/optimizer/JSDCE-output.js
+++ b/test/optimizer/JSDCE-output.js
@@ -14,6 +14,19 @@ function g(a) {
 
 Module["g"] = g;
 
+const sx = {
+ a: 1
+};
+
+const sar = [ 1, 2, 3 ];
+
+Module["spread"] = {
+ b: 2,
+ ...sx
+};
+
+Module["spread2"] = [ ...sar, 4, 5, 6 ];
+
 function h(a) {
  return a + 1;
 }
```

File: test/optimizer/JSDCE.js [hunk 1]
```diff
@@ -19,6 +19,10 @@ function g(a) {
   return a+1;
 }
 Module['g'] = g;
+const sx = {a:1};
+const sar = [1,2,3];
+Module['spread'] = { b:2, ...sx };
+Module['spread2'] = [ ...sar, 4, 5, 6 ];
```
Comment (kripken) [inline test/optimizer/JSDCE.js:25]:
```suggestion
const sx = {
 a: 1
};

const sar = [ 1, 2, 3 ];

Module["spread"] = {
 b: 2,
 ...sx
};

Module["spread2"] = [ ...sar, 4, 5, 6 ];
```

File: test/optimizer/JSDCE-output.js [hunk 2]
```diff
@@ -14,6 +14,14 @@ function g(a) {
 
 Module["g"] = g;
 
+const sx = {a:1};
+
+const sar = [1,2,3];
+
+Module['spread'] = { b:2, ...sx };
+
+Module['spread2'] = [ ...sar, 4, 5, 6 ];
```
Comment (kripken) [inline test/optimizer/JSDCE-output.js:23]:
```suggestion
const sx = {
 a: 1
};

const sar = [ 1, 2, 3 ];

Module["spread"] = {
 b: 2,
 ...sx
};

Module["spread2"] = [ ...sar, 4, 5, 6 ];
```

**Comment by kripken** [comment]
Committed, should pass I hope...

**Pull request merged**